### PR TITLE
libmbedtls: mbedtls_mpi_exp_mod(): reduce stack usage

### PR DESCRIPTION
**Pull request by jenswi-linaro**
The W variable is 3072 on AArch64 with MBEDTLS_MPI_WINDOW_SIZE set to 6
for maximum performance. Instead of allocating such a large variable on
the stack use mempool_alloc().

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jforissier** [comment]
"is 3072 **bytes**". Other than that, LGTM.

**Comment by jenswi-linaro** [comment]
Thanks, I'll fix it when I apply a tag. Do I have your A-B or R-B?

**Comment by jenswi-linaro** [comment]
IBART is probably unhappy because the `mbedtls-2.16.0` branch is a bit old. If we want IBART to succeed here we'll need to merge `master` into `mbedtls-2.16.0` and rebase this PR.

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

> If we want IBART to succeed here we'll need to merge `master` into `mbedtls-2.16.0`

Agreed, done.

**Comment by jenswi-linaro** [comment]
Rebased, tag applied and commit message updated.

**Comment by jforissier** [comment]
IBART is still failing, something must be wrong?

File: lib/libmbedtls/mbedtls/library/bignum.c [hunk 1]
```diff
@@ -1802,13 +1804,21 @@ int mbedtls_mpi_exp_mod( mbedtls_mpi *X, const mbedtls_mpi *A,
     if( mbedtls_mpi_cmp_int( E, 0 ) < 0 )
         return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
 
+    W = mempool_alloc( mbedtls_mpi_mempool,
+                       sizeof( mbedtls_mpi ) * array_size_W );
+    if( W == NULL )
+    {
+        ret = MBEDTLS_ERR_MPI_ALLOC_FAILED;
```
Comment (etienne-lms) [inline lib/libmbedtls/mbedtls/library/bignum.c]: Prefer `return MBEDTLS_ERR_MPI_ALLOC_FAILED;` straight otherwise `mbedtls_mpi_free()` is called over uninited `T` and `Apos`.

Comment (jenswi-linaro) [reply lib/libmbedtls/mbedtls/library/bignum.c]: Good catch, I'll fix.

**Comment by jenswi-linaro** [comment]
It seems the memory pool used by TAs needs to be adjusted, I'll try something.

File: lib/libmbedtls/mbedtls/library/bignum.c [hunk 2]
```diff
@@ -1802,13 +1804,18 @@ int mbedtls_mpi_exp_mod( mbedtls_mpi *X, const mbedtls_mpi *A,
     if( mbedtls_mpi_cmp_int( E, 0 ) < 0 )
         return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
 
+    W = mempool_alloc( mbedtls_mpi_mempool,
+                       sizeof( mbedtls_mpi ) * array_size_W );
+    if( W == NULL )
+        return( MBEDTLS_ERR_MPI_ALLOC_FAILED);
```
Comment (etienne-lms) [inline lib/libmbedtls/mbedtls/library/bignum.c]: minor: space before `)` :)

Comment (jenswi-linaro) [reply lib/libmbedtls/mbedtls/library/bignum.c]: I'll fix.
That coding style is hard to follow sometimes, good thing you seem to have an eye for it. :-)

**Comment by jenswi-linaro** [comment]
IBART was happy with "libutee: increase MPI mempool size" commit.
I've fixed up "libmbedtls: mbedtls_mpi_exp_mod(): reduce stack usage" and applied Etienne's tag too.
I just realized after pushing that I'm missing tags for "libutee: increase MPI mempool size"

**Comment by jforissier** [comment]
Minor: perhaps the commit text in "libutee: increase MPI mempool size" should cite the (future) commit "libmbedtls: mbedtls_mpi_exp_mod(): reduce stack usage", I mean near "compensate for changes...", this is the change we are talking about, it is not a fix for a prior change.

**Comment by jenswi-linaro** [comment]
Updated the commit message in "libutee: increase MPI mempool size". I guess you've noted that "libutee: increase MPI mempool size" comes first even if github is trying to fool us.

**Comment by jforissier** [comment]
> I guess you've noted that "libutee: increase MPI mempool size" comes first even if github is trying to fool us.

Uh, no, I've been fooled again :) Actually I would argue it should **not** come first to avoid breaking bisection.

**Comment by jenswi-linaro** [comment]
It is "libutee: increase MPI mempool size" that we want first for bisection. I guess you misread or something. :-)

**Comment by jforissier** [comment]
@jenswi-linaro yeah ignore my comment, I need more coffee :) all is fine with me now :+1:

**Comment by jforissier** [comment]
...and BTW: for "libutee: increase MPI mempool size "
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied
